### PR TITLE
Prevent search query being reset when changing object type

### PR DESCRIPTION
**Pull request by mk-simsoft**
The search input is no longer reset when the object changes or clicks outside of input.

Closes [AB#691](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/691), [AB#830](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/830)

#### User changes
- The search input is no longer reset when the object changes.

#### Superuser changes
- None.

#### Admin changes
- None.

#### System admin changes
- [x] anet.yml or anet-dictionary.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
  - [x] Described the user behavior in PR body
  - [x] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [ ] Added and/or updated unit tests
  - [x] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [ ] Opened debt issues for anything not resolved here

### CODE REVIEW

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/components/advancedSelectWidget/MultiTypeAdvancedSelectComponent.js [hunk 1]
```diff
@@ -214,6 +214,7 @@ const MultiTypeAdvancedSelectComponent = ({
         placeholder={searchPlaceholder}
         value={value}
         showEmbedded
+        reset={true}
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/MultiTypeAdvancedSelectComponent.js:217]: A boolean prop like this with value `true` can simply be specified as `reset` (there's a lint rule for this [`react/jsx-boolean-value`], but it's not enabled). However, the issue only applies when the user has multiple entityTypes to choose from, so the value should probably be derived from that, or passed down from the place where the component is used. In any case, the current code changes don't achieve the intended goal, since when I type e.g. "ANET", then change the entityType From Reports to People, the search text stays there, but no search results are shown anymore, and as soon as I click in the input field to fire the search, the search text is cleared.

File: client/src/components/advancedSelectWidget/AdvancedSelect.js [hunk 2]
```diff
@@ -275,19 +277,20 @@ const AdvancedSelect = ({
...
+      if (!reset) {
+        setSearchTerms(selectedValueAsString)
+      }
       setResults({})
       setPageNum(0)
       setFetchType(FETCH_TYPE.NONE)
       setDoReset(false)
     }
-  }, [doReset, firstFilter, selectedValueAsString])
+  }, [doReset, firstFilter, selectedValueAsString, reset])
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/AdvancedSelect.js:293]: This is confusing; we now have similarly named vars `doReset` and `reset` that mean different things; `reset` should probably be renamed to something clearer (e.g. `applySearchTerms` or so).

**Review by gjvoosten — CHANGES_REQUESTED** [review]
A change like this could do with some client-side tests to verify that the current implementation fails the requirement, and that the new (intended) implementation meets it.

**Comment by midmarch** [comment]
Looks like one test is failing @mk-simsoft

File: client/src/components/advancedSelectWidget/AdvancedSelect.js [hunk 3]
```diff
@@ -440,6 +444,7 @@ const AdvancedSelect = ({
 
   function changeSearchTerms(event) {
     setSearchTerms(event.target.value)
+    console.log(searchTerms)
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/AdvancedSelect.js:447]: Remove left-over debugging.

File: client/src/components/advancedSelectWidget/AdvancedSelect.js [hunk 4]
```diff
@@ -275,20 +277,20 @@ const AdvancedSelect = ({
...
   useEffect(() => {
     if (doReset) {
       setIsLoading(false)
       setShowOverlay(false)
       setFilterType(firstFilter)
-      setSearchTerms(selectedValueAsString)
+      if (!applySearchTerms) {
+        setSearchTerms(selectedValueAsString)
+      }
       setResults({})
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/AdvancedSelect.js:288]: This seems counterintuitive: if it does not apply, set it. I would expect: "if it does apply, set it", i.e. the opposite of the current condition.

File: client/src/components/advancedSelectWidget/AdvancedSelect.js [hunk 5]
```diff
@@ -372,7 +374,7 @@ const AdvancedSelect = ({
                 <InputGroup>
                   <Form.Control
                     name={fieldName}
-                    value={searchTerms || ""}
+                    value={searchTerms}
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/AdvancedSelect.js:377]: Revert this change; if `searchTerms` is `undefined` or `null`, the form control will have no value (which is an error, and `undefined` will even make it an uncontrolled input which you can't give a value).

**Review by gjvoosten — CHANGES_REQUESTED** [review]
Solution doesn't work for e.g. **Related report** (in report form, custom field under **Engagement information**) where it doesn't reset the input like it does in other places where a single entity type is shown, or **Additional engagement needed for** where it doesn't remember the search input when changing the entity type.

File: client/src/components/editor/LinkSourceAnet.js [hunk 6]
```diff
@@ -66,6 +66,7 @@ const LinkSourceAnet = ({
               const anetLinkNode = createAnetLinkNode(objectType, value.uuid)
               insertAnetLink(anetLinkNode)
             }}
+            keepSearchText
```
Comment (gjvoosten) [inline client/src/components/editor/LinkSourceAnet.js:69]: Solution doesn't work for e.g. **Additional engagement needed for** (in report form, custom field under Engagement information) and in the related objects of a note: in both places it doesn't remember the search input when changing the entity type.
Additionally, a change like this could do with some client-side tests to verify that the current implementation fails the requirement, and that the new (intended) implementation meets it.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/components/editor/LinkSourceAnet.js (hunk 6 above)
Comment (hyesilAnka) [reply client/src/components/editor/LinkSourceAnet.js:69]: Actually, the main intent was to support this functionality in rich text-editor. Therefore, of course it will not work for other components. After adding new capability to remember search query, for now it has been called only in Rich-Text-editor.  To support note related object I will apply new changes but I am not sure for the report page,  may be we should discuss with Maarteen.

File: anet-dictionary.yml [hunk 7]
```diff
@@ -302,6 +302,17 @@ fields:
...
+        type: anet_object
+        label: Related object
+        helpText: Here you can link to a related object
+        types:
+          - Report
+          - Person
+          - Position
+          - Organization
+          - Location
+          - Task
```
Comment (gjvoosten) [inline anet-dictionary.yml:315]: Don't forget to add this same change to `testDictionaries/no-custom-fields.yml`.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: anet-dictionary.yml (hunk 7 above)
Comment (hyesilAnka) [reply anet-dictionary.yml:315]: Thanks for reminding, added.  In  fact, it looks like I was the first to put the `customfields` property for the report into the test dictionary. 
Like, it'i never been done before.

Comment (gjvoosten) [reply anet-dictionary.yml:315]: Ah, my bad, the test dictionary should be identical to the main one, with the exception that it should contain _no_ customFields at all. So in fact you shouldn't add that change to the test dictionary. Sorry about that… :blush:

Comment (hyesilAnka) [reply anet-dictionary.yml:315]: 😄 Ok, taking it back.

**Review by gjvoosten — APPROVED** [review]
